### PR TITLE
cbqn: init at 0.0.0+unstable=2021-09-26

### PR DESCRIPTION
**Pull request by AndersonTorres**
###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/#sec-conf-file))
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by sternenseemann** [comment]
Arguably #135665 is the more elegant solution as it allows for flexible bootstrapping as well as “true” self-bootstrapping.

**Comment by AndersonTorres** [comment]
More or less.

The original JS-BQN should be build first (as a Node application).

File: pkgs/development/interpreters/bqn/cbqn/default.nix [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+
+stdenv.mkDerivation rec {
+  pname = "cbqn";
+  version = "0.0.0+unstable=2021-09-29";
+
+  src = fetchFromGitHub {
+    owner = "dzaima";
+    repo = "CBQN";
+    rev = "1c83483d5395e097f60de299274ebe0df590217e";
+    hash = "sha256-C34DpXab08mBm2oCQuaeq4fJPtQ5rVa/HlpL/nB9XjQ=";
```
Comment (SuperSandro2000) [inline pkgs/development/interpreters/bqn/cbqn/default.nix:15]:
```suggestion
    sha256 = "sha256-C34DpXab08mBm2oCQuaeq4fJPtQ5rVa/HlpL/nB9XjQ=";
```

File: pkgs/development/interpreters/bqn/cbqn/default.nix [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+    then ''
+      cp ${cbqn-bytecode}/src/gen/{compiler,formatter,runtime0,runtime1,src} src/gen/
+    ''
+    else ''
+      ${bqn-path} genRuntime ${mlochbaum-bqn}
+    '';
+
+  makeFlags = [
+    "CC=${stdenv.cc.targetPrefix}cc"
+    "single-c"
```
Comment (jshholland) [inline pkgs/development/interpreters/bqn/cbqn/default.nix:54]: Is there a reason not to be using one of `single-o3` or `single-o3g` to have an optimised build here?

File: pkgs/development/interpreters/bqn/cbqn/default.nix [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+  ];
+
+  installPhase = ''
+     runHook preInstall
+
+     mkdir -p $out/bin/
+     cp BQN -t $out/bin/
+     ln -s $out/bin/BQN $out/bin/bqn
```
[the remark this answers is not in the document]
Comment (AndersonTorres) [reply pkgs/development/interpreters/bqn/cbqn/default.nix:62]: There are many tutorials on the internet that employ both.

File: pkgs/development/interpreters/bqn/cbqn/default.nix (hunk 2 above)
Comment (AndersonTorres) [reply pkgs/development/interpreters/bqn/cbqn/default.nix:54]: No. I just tried the easiest one.

File: pkgs/top-level/all-packages.nix [hunk 4]
```diff
@@ -12789,6 +12789,19 @@ with pkgs;
...
+  # And the classic bootstrapping process
+  cbqn-phase0 = callPackage ../development/interpreters/bqn/cbqn {
+    bqn-path = null;
+  };
+  cbqn-phase1 = callPackage ../development/interpreters/bqn/cbqn {
+    bqn-path = "${cbqn-phase0}/bin/bqn";
+  };
+  cbqn-phase2 = callPackage ../development/interpreters/bqn/cbqn {
+    bqn-path = "${cbqn-phase1}/bin/bqn";
+  };
```
[the remark this answers is not in the document]
Comment (AndersonTorres) [reply pkgs/top-level/all-packages.nix:12803]: How exactly this can be made?

**Pull request merged**